### ISSUE TITLE
Performance regression when ZONE_ECHO_LIMIT changed from 7 to 100

### ISSUE THREAD

**Issue by nifgraup** [dfahlander/Dexie.js#1161]
Regression point 3046e3684cb6092d31c634e36521364070a9f87b, #722

I don't have reproduction steps but thought I would share findings so far.

**Comment by dfahlander** [comment]
The ZONE_ECHO_LIMIT is only hit if on one of the following rare scenarios occur:

* A) A promise other than window.Promise (which is temprorary equal to Dexie.Promise within transactions zone) is **await**ed (natively) within a transaction. One example is awaiting a native promise from the crypto API. This will not work anyway as IndexedDB transactions won't survive it.
* B) The code awaits several non-promise values after each other:
  ```js
  db.transaction('rw', db.friends, async tx => {
     await "non-promise1";
     await "non-promise2";
     ...
     await "non-promise100";
  });  ...
  ```
  *For an example of this situation, see issue #774*

So in conclusion, the increase of this limit does not degrade performance in normal apps.

*NOTE: There's also [another commit](https://github.com/dfahlander/Dexie.js/commit/d02d999858eb9a449ce198b391db8849269d2cdb) that reverts part of this commit, but not the increase of MAX_ECHO_LIMIT.*

**Comment by nifgraup** [comment]
Thanks for the explaination. The only Promises I am **await**ing on in transactions, other than window.Promise, are async functions. They return a native Promise when there is no return statement or if a conditional return is not hit.

I can rewrite this to use window.Promise but I would need some way to prevent the performance degration from being reintroduced as this is quite subtle. Perhaps by being able to set ZONE_ECHO_LIMIT to 0 or 1. That's how I discovered what was going on in the first place.

**Comment by dfahlander** [comment]
There should be no problem to call async functions that either return a direct value (non-promise) or return a promise originated from a Dexie.js call. The called async function can call other async functions as well. The only thing to avoid is calling non-dexie calls from within transactions. If you still find that the the ZONE_ECHO_LIMIT is being hit despite following this guideline, I would be keen on finding a repro for that if possible.

**Comment by nifgraup** [comment]
I can hit the limit this way:
```
-var ZONE_ECHO_LIMIT = 100;
+var ZONE_ECHO_LIMIT = 1;
```
then executing
```
async function waitMe(): Promise<void> {
  await Promise.resolve();
}
db.transaction("rw", [db.tableName], async () => {
  console.log(Promise.resolve().then+'');
  await waitMe();
  console.log(Promise.resolve().then+'');
});
```
The first console.log is a Dexie Promise while the second is native Promise.

**Comment by dfahlander** [comment]
Ok, but that is natural since you've lowered it to 1, which will always be hit. The question is that if it is a risk that performance will suffer due to the increased limit of 100, so we must reproduce a way to hit that limit.

**Comment by nifgraup** [comment]
Right, here is the reproduction:
```
db.transaction("rw", [db.tableName], () => {
  dexie.Promise.all(thousandEntries.map(async (entry) => {
    const existing = await db.tableName.where("foo").equals(entry.foo).toArray();
    if (existing.length > 0) {
      // do stuff
      db.tableName.put(entry);
    } else {
      db.tableName.add(entry);
    }
  }));
  // something that doesn't depend on the block above but needs the transaction
  // db.tableName.add(...)
});
```

**Comment by dfahlander** [comment]
I don't see how this code even enters the state of using zone echoing as the callback to db.transaction is not async.
And if you had it async, the only thing I expect to happen here is that the Promise.all() will make sure to keep the zones within all the thousands promises (which is needed and not because of the limit).

Have you ran this code and printed out what is happening?

**Comment by nifgraup** [comment]
True, it a was transaction callback with async. I'll add console logs.

**Comment by dfahlander** [comment]
Just a note: Promise.all() will spawn as many zone echoes as there are items in the given array no matter the limit. So if the given array contains 1000 items, you will see 1000 echoes. This is how the zone echoing need to function in case every item in the array is a non-promise or an async function that each one awaits up to ZONE_ECHO_LIMIT number of non-promises. In your sample it does only await promises originating from dexie calls, so I would expect 1000 zone echoes to take place there.

**Comment by nifgraup** [comment]
I'm not able to see the limit break with console logs in the loop as I did here https://github.com/dfahlander/Dexie.js/issues/1161#issuecomment-720383732 .

But do see the performance issue go away by either setting `ZONE_ECHO_LIMIT` back to 7 (or lower) or change the loop to
```
  dexie.Promise.all(thousandEntries.map((entry) => {
    return db.tableName.where("foo").equals(entry.foo).toArray().then(existing => {
      if (existing.length > 0) {
        // do stuff
        db.tableName.put(entry);
      } else {
        db.tableName.add(entry);
      }
    });
  }));
```

**Comment by dfahlander** [comment]
How does the performance issue manifest?
What is the reason you use Promise.all() without awaiting it? Would you experience a performance improvement if you keep it as an async transaction but `await` Promise.all()? If you're not awaiting it you could equally well replace the `Promise.all(thousandEntries.map(...))` with just `thousandEntries.forEach(...)`.

**Comment by nifgraup** [comment]
It's a 5 sec vs. 2 sec run time with a test case I have in the original code.
The reason for not awaiting Promise.all is that there is code that following it that has to be in the transaction as well but can be executed right away as it doesn't depend on the code in the Promise.all. Awaiting on Promise.all does not speed up the code.

Dexie.Promise.all is used here to make sure all the promises are resolve within the transaction. I do get an error `TransactionInactiveError: Transaction has already completed or failed` if I change to `thousandEntries.forEach(async (entry) => ...`

**Comment by dfahlander** [comment]
I'm trying to reproduce this. Could you help me fork the following jsitor fiddle: https://jsitor.com/ZicMSTbmc and try reproduce the problem. In the HTML tab you could switch between dexie 3.0.2 and an older version that uses a lower ZONE_ECHO_LIMIT (like 2.0.4).

**Comment by nifgraup** [comment]
Sorry about the delay, here is a proper repro https://jsitor.com/7DN7DZkLh

**Comment by dfahlander** [comment]
Thanks a ton for the repro!
I've found the root problem that can get into unbalance in zone echoing and made a fix for it. The problem occurs because we await dexie promises without having entered the native async mode - which made it enter that mode dynamically without scheduling a cleanup when leaving it.

**Comment by nifgraup** [comment]
Welcome. Your fix takes care of vast majority of the problem. You can still see a difference between ZONE_ECHO_LIMIT 7 and 100 in the test case above, not sure if something can be done about that.

**Comment by dfahlander** [comment]
The remaining difference is because the first measure also opens the database as well for JIT optimizations taking place the in the first rum. If you add a line await db.open() before the first measuere - or switch order between the two you will find that there is no difference. You can even increase ZONE_ECHO_LIMIT to 500 or 100000 and you will still see no difference.

https://jsitor.com/AJSuC8ynv
